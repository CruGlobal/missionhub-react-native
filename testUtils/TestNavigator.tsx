import React, { Children } from 'react';
import {
  createAppContainer,
  createSwitchNavigator,
  NavigationComponent,
  NavigationParams,
} from 'react-navigation';

export const TestNavigator = ({
  children,
  params,
}: {
  children: NavigationComponent;
  params?: NavigationParams;
}) => {
  const Navigator = createAppContainer(
    createSwitchNavigator({
      TestScreen: { screen: () => Children.only(children), params },
    }),
  );
  return <Navigator />;
};
