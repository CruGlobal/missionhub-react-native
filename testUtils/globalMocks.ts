import faker from 'faker/locale/en';
import { IMocks } from 'graphql-tools';
import moment from 'moment';

import {
  CommunityCelebrationCelebrateableEnum,
  PermissionEnum,
  ReminderTypeEnum,
  StepTypeEnum,
} from '../__generated__/globalTypes';

let currentId = 1;
const nextId = () => currentId++;
export const resetGlobalMockSeeds = () => {
  currentId = 1;
  faker.seed(1);
};
resetGlobalMockSeeds();

export const globalMocks: IMocks = {
  String: () => faker.lorem.words(),
  Int: () => faker.random.number(),
  Float: () => faker.random.number({ precision: 0.01 }),
  Boolean: () => faker.random.boolean(),
  ID: () => nextId(),
  ISO8601DateTime: () => faker.date.past(10, '2020-01-14').toISOString(),
  ISO8601Date: () =>
    moment(faker.date.past(10, '2020-01-14')).format('YYYY-MM-DD'),

  BasePageInfo: () => ({
    endCursor: null,
    hasNextPage: false,
    hasPreviousPage: false,
    startCursor: null,
  }),
  ReminderTypeEnum: () => ReminderTypeEnum.once,
  StepTypeEnum: () => faker.random.arrayElement(Object.values(StepTypeEnum)),

  Step: () => ({
    title: faker.lorem.sentence(),
  }),
  Person: () => {
    const firstName = faker.name.firstName();
    const lastName = faker.name.lastName();
    return {
      fullName: `${firstName} ${lastName}`,
      firstName,
      lastName,
    };
  },
  Community: () => ({
    name: faker.company.catchPhrase(),
  }),
  CommunityCelebrationItem: () => {
    const firstName = faker.name.firstName();
    const lastName = faker.name.lastName();
    return {
      celebrateableType: faker.random.arrayElement(
        Object.values(CommunityCelebrationCelebrateableEnum),
      ),
      changedAttributeValue: moment(
        faker.date.past(10, '2020-01-14'),
      ).toISOString(),
      subjectPersonName: `${firstName} ${lastName}`,
    };
  },
  CommunityPermission: () => {
    return {
      permission: faker.random.arrayElement(Object.values(PermissionEnum)),
    };
  },
  FeedItem: () => {
    const firstName = faker.name.firstName();
    const lastName = faker.name.lastName();
    return {
      subjectPersonName: `${firstName} ${lastName}`,
    };
  },
};
