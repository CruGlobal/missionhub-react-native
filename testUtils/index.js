import 'react-native';
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

// Need to import for all translation screens to work properly
import '../src/i18n';

Enzyme.configure({ adapter: new Adapter() });

export const createMockStore = (state = {}) => {
  return {
    getState: jest.fn(() => (state)),
    dispatch: jest.fn((response) => Promise.resolve(response)),
    subscribe: jest.fn(),
  };
};

export const createMockNavState = (params = {}) => {
  return { state: { params } };
};

export const testSnapshot = (data) => {
  expect(renderer.create(data)).toMatchSnapshot();
};

export const renderShallow = (component, store) => {
  let renderedComponent = shallow(
    component,
    { context: { store: store } }
  );

  // If component has translation wrappers, dive deeper
  while (renderedComponent.is('Translate') || renderedComponent.is('I18n')) {
    renderedComponent = renderedComponent.dive();
  }

  // Render contents of component
  renderedComponent = renderedComponent.dive();
  return renderedComponent;
};

export const testSnapshotShallow = (component, store) => {
  const renderedComponent = renderShallow(component, store);
  expect(renderedComponent).toMatchSnapshot();
  return renderedComponent;
};

export const mockFnWithParams = (obj, method, expectedReturn, ...expectedParams) => {
  return obj[method] = jest.fn(
    (...actualParams) => {
      expect(actualParams).toEqual(expectedParams);
      return expectedReturn;
    }
  );
};
