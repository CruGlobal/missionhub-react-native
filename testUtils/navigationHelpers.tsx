import React from 'react';
import {
  // eslint-disable-next-line import/named
  createAppContainer,
  // eslint-disable-next-line import/named
  createSwitchNavigator,
  // eslint-disable-next-line import/named
  NavigationParams,
  // eslint-disable-next-line import/named
  NavigationScreenProp,
} from 'react-navigation';
import { render } from 'react-native-testing-library';

export const createNavigationProp = (params?: NavigationParams) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let navigationProp: NavigationScreenProp<any> | undefined;

  const Navigator = createAppContainer(
    createSwitchNavigator({
      TestScreen: {
        screen: ({
          navigation,
        }: {
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          navigation: NavigationScreenProp<any>;
        }) => {
          navigationProp = navigation;
          return null;
        },
        params,
      },
    }),
  );
  render(<Navigator />);

  if (!navigationProp) {
    throw 'Unable to get navigation screen prop';
  }

  return navigationProp;
};
