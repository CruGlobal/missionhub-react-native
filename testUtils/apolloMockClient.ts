import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';
import { addMockFunctionsToSchema, IMocks } from 'graphql-tools';
import { DocumentNode } from 'graphql';
import { buildClientSchema, IntrospectionQuery } from 'graphql/utilities';
import gql from 'fraql';
import { createMockerFromIntrospection } from 'fraql/mock';

import introspectionQuery from '../schema.json';

import { globalMocks } from './globalMocks';

export const createApolloMockClient = (mocks: IMocks = {}) => {
  const schema = buildClientSchema(
    (introspectionQuery as unknown) as IntrospectionQuery,
  );

  addMockFunctionsToSchema({
    schema,
    mocks: { ...globalMocks, ...mocks },
  });

  return new ApolloClient({
    cache: new InMemoryCache(),
    link: new SchemaLink({ schema }),
  });
};

export const mockFragment = <TData>(
  fragmentDocument: DocumentNode,
  options?: {
    mocks: IMocks;
  },
): TData => {
  if (!fragmentDocument.loc) {
    throw 'Unable to get original gql query to create FraQL fragment.';
  }

  return createMockerFromIntrospection(introspectionQuery, {
    mocks: globalMocks,
  }).mockFragment(gql(fragmentDocument.loc.source.body), options);
};
