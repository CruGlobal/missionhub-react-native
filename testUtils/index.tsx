import React, { ReactElement, ReactNode } from 'react';
import 'react-native';
import { Provider } from 'react-redux';
import { Provider as ProviderLegacy } from 'react-redux-legacy';
import thunk from 'redux-thunk';
import configureStore, { MockStore } from 'redux-mock-store';
// eslint-disable-next-line import/named
import { NavigationParams, NavigationProvider } from 'react-navigation';
import { ApolloProvider } from '@apollo/react-hooks';
import { ReactTestRendererJSON } from 'react-test-renderer';
import { render, RenderAPI } from 'react-native-testing-library';
import snapshotDiff from 'snapshot-diff';
import Enzyme, { shallow as enzymeShallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { IMocks } from 'graphql-tools';

import { createApolloMockClient } from './apolloMockClient';
import { createNavigationProp } from './navigationHelpers';

Enzyme.configure({ adapter: new Adapter() });

export const createThunkStore = configureStore([thunk]);

interface RenderWithContextParams {
  initialState?: {};
  store?: MockStore;
  navParams?: NavigationParams;
  mocks?: IMocks;
  noWrappers?: boolean;
}

// Inspiration from https://github.com/kentcdodds/react-testing-library/blob/52575005579307bcfbe7fbe4ef4636147c03c6fb/examples/__tests__/react-redux.js#L69-L80
export function renderWithContext(
  component: ReactElement,
  {
    initialState,
    store = createThunkStore(initialState),
    navParams,
    mocks: mocks = {},
    noWrappers = false,
  }: RenderWithContextParams = {},
) {
  const mockApolloClient = createApolloMockClient(mocks);

  const navigation = createNavigationProp(navParams);

  let renderResult: RenderAPI;
  if (noWrappers) {
    renderResult = render(React.cloneElement(component, { navigation }));
  } else {
    // Warning: don't call any functions in here that return new instances on every call. All the props need to stay the same otherwise rerender won't work.
    const wrapper = ({ children }: { children: ReactNode }) => (
      <NavigationProvider value={navigation}>
        <ProviderLegacy store={store}>
          <Provider store={store}>
            <ApolloProvider client={mockApolloClient}>
              {children}
            </ApolloProvider>
          </Provider>
        </ProviderLegacy>
      </NavigationProvider>
    );

    renderResult = render(React.cloneElement(component, { navigation }), {
      wrapper,
    });
  }

  let storedSnapshot: ReactTestRendererJSON | null;
  return {
    ...renderResult,
    store,
    snapshot: () => expect(renderResult.toJSON()).toMatchSnapshot(),
    recordSnapshot: () => (storedSnapshot = renderResult.toJSON()),
    diffSnapshot: () => {
      if (!storedSnapshot) {
        throw new Error(
          'You must call recordSnapshot to store an initial snapshot before calling diffSnapshot',
        );
      }
      expect(
        snapshotDiff(storedSnapshot, renderResult.toJSON()),
      ).toMatchSnapshot();
    },
  };
}

// TODO: Remove all legacy rendering functions below

export const createMockNavState = (params = {}) => {
  return { state: { params } };
};

export const testSnapshot = (component: ReactElement) => {
  const { toJSON } = render(component);
  expect(toJSON()).toMatchSnapshot();
};

export const renderShallow = (
  component: ReactElement,
  store = createThunkStore(),
) => {
  let renderedComponent = enzymeShallow(
    <ProviderLegacy store={store}>{component}</ProviderLegacy>,
  ).dive();

  // If component has translation wrappers, dive deeper
  while (
    // eslint-disable-next-line  @typescript-eslint/no-explicit-any
    ((renderedComponent.type() as any).displayName || '').startsWith(
      'withI18nextTranslation(',
    )
  ) {
    renderedComponent = renderedComponent.dive();
  }

  // Render contents of component
  renderedComponent = renderedComponent.dive();
  return renderedComponent;
};

export const testSnapshotShallow = (
  component: ReactElement,
  store = createThunkStore(),
) => {
  const renderedComponent = renderShallow(component, store);
  expect(renderedComponent).toMatchSnapshot();
  return renderedComponent;
};
