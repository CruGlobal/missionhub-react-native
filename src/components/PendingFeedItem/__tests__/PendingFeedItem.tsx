import React from 'react';
import { fireEvent } from 'react-native-testing-library';

import {
  PostTypeEnum,
  CreatePostInput,
  UpdatePostInput,
} from '../../../../__generated__/globalTypes';
import {
  StoredCreatePost,
  StoredUpdatePost,
} from '../../../reducers/communityPosts';
import {
  deletePendingPost,
  useCreatePost,
  useUpdatePost,
} from '../../../containers/Groups/CreatePostScreen';
import { renderWithContext } from '../../../../testUtils';
import { PendingFeedItem } from '../';

jest.mock('react-native-video', () => 'Video');
jest.mock('../../../containers/Groups/CreatePostScreen');

const pendingItemId = '1';
const media = 'asdf.mov';
const content = 'text';
const communityId = '1234';
const postType = PostTypeEnum.story;
const postId = '2';

const onComplete = jest.fn();
const createPost = jest.fn();
const updatePost = jest.fn();

const storedCreatePost: StoredCreatePost = {
  storageId: pendingItemId,
  media,
  content,
  communityId,
  postType,
  failed: false,
};
const storedUpdatePost: StoredUpdatePost = {
  id: postId,
  storageId: pendingItemId,
  media,
  content,
  communityId,
  failed: false,
};
const createPostInput: CreatePostInput = {
  communityId,
  content,
  postType,
  media,
};
const updatePostInput: UpdatePostInput = {
  id: postId,
  content,
  media,
};

const deletePendingPostResponse = { type: 'delete post' };

beforeEach(() => {
  (useCreatePost as jest.Mock).mockReturnValue(createPost);
  (useUpdatePost as jest.Mock).mockReturnValue(updatePost);
  (deletePendingPost as jest.Mock).mockReturnValue(deletePendingPostResponse);
});

describe('PendingFeedItem', () => {
  it('renders pending post', () => {
    renderWithContext(
      <PendingFeedItem pendingItemId={pendingItemId} onComplete={onComplete} />,
      {
        initialState: {
          communityPosts: {
            pendingPosts: {
              [pendingItemId]: storedCreatePost,
            },
          },
        },
      },
    ).snapshot();
  });

  it('renders failed post', () => {
    renderWithContext(
      <PendingFeedItem pendingItemId={pendingItemId} onComplete={onComplete} />,
      {
        initialState: {
          communityPosts: {
            pendingPosts: {
              [pendingItemId]: {
                ...storedCreatePost,
                failed: true,
              },
            },
          },
        },
      },
    ).snapshot();
  });

  it('cancels post upload', () => {
    const { getByTestId } = renderWithContext(
      <PendingFeedItem pendingItemId={pendingItemId} onComplete={onComplete} />,
      {
        initialState: {
          communityPosts: {
            pendingPosts: {
              [pendingItemId]: {
                ...storedCreatePost,
                failed: true,
              },
            },
          },
        },
      },
    );

    fireEvent.press(getByTestId('CancelButton'));

    expect(deletePendingPost).toHaveBeenCalledWith(pendingItemId);
  });

  it('retries post create', () => {
    const { getByTestId } = renderWithContext(
      <PendingFeedItem pendingItemId={pendingItemId} onComplete={onComplete} />,
      {
        initialState: {
          communityPosts: {
            pendingPosts: {
              [pendingItemId]: {
                ...storedCreatePost,
                failed: true,
              },
            },
          },
        },
      },
    );

    fireEvent.press(getByTestId('RetryButton'));

    expect(createPost).toHaveBeenCalledWith(createPostInput);
  });

  it('retries post update', () => {
    const { getByTestId } = renderWithContext(
      <PendingFeedItem pendingItemId={pendingItemId} onComplete={onComplete} />,
      {
        initialState: {
          communityPosts: {
            pendingPosts: {
              [pendingItemId]: {
                ...storedUpdatePost,
                failed: true,
              },
            },
          },
        },
      },
    );

    fireEvent.press(getByTestId('RetryButton'));

    expect(updatePost).toHaveBeenCalledWith(updatePostInput, communityId);
  });
});
