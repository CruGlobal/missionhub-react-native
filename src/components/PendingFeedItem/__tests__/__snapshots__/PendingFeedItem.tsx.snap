// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PendingFeedItem renders failed post 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#ffffff",
        "borderRadius": 8,
        "elevation": 4,
        "flexDirection": "row",
        "marginHorizontal": 16,
        "marginVertical": 8,
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.3,
        "shadowRadius": 2,
      },
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "flex-start",
        "padding": 16,
      },
    ]
  }
  testID="PendingFeedItem"
>
  <Video
    controls={false}
    paused={true}
    resizeMode="cover"
    source={
      Object {
        "uri": "",
      }
    }
    style={
      Object {
        "height": 48,
        "width": 48,
      }
    }
  />
  <View
    style={
      Object {
        "flex": 1,
        "marginHorizontal": 7,
      }
    }
  >
    <View>
      <Text
        style={
          Object {
            "color": "#505256",
            "fontFamily": "SourceSansPro-Regular",
            "fontSize": 16,
            "fontStyle": "normal",
            "lineHeight": 24,
            "textAlignVertical": "center",
            "textTransform": "none",
          }
        }
      >
        Failed to upload video.
      </Text>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="RetryButton"
      >
        <Text
          style={
            Object {
              "color": "#3CC8E6",
              "fontFamily": "SourceSansPro-Bold",
              "fontSize": 16,
              "fontStyle": "normal",
              "lineHeight": 24,
              "textAlignVertical": "center",
              "textTransform": "none",
            }
          }
        >
          Try again.
        </Text>
      </View>
    </View>
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="CancelButton"
  >
    <SvgMock
      color="#B4B6BA"
    />
  </View>
</View>
`;

exports[`PendingFeedItem renders pending post 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#ffffff",
        "borderRadius": 8,
        "elevation": 4,
        "flexDirection": "row",
        "marginHorizontal": 16,
        "marginVertical": 8,
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.3,
        "shadowRadius": 2,
      },
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "flex-start",
        "padding": 16,
      },
    ]
  }
  testID="PendingFeedItem"
>
  <Video
    controls={false}
    paused={true}
    resizeMode="cover"
    source={
      Object {
        "uri": "",
      }
    }
    style={
      Object {
        "height": 48,
        "width": 48,
      }
    }
  />
  <View
    style={
      Object {
        "flex": 1,
        "marginHorizontal": 7,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#505256",
          "fontFamily": "SourceSansPro-Regular",
          "fontSize": 16,
          "fontStyle": "normal",
          "lineHeight": 24,
          "textAlignVertical": "center",
          "textTransform": "none",
        }
      }
    >
      Posting...
    </Text>
  </View>
  <ActivityIndicator
    animating={true}
    color="rgba(0, 0, 0, 1)"
    hidesWhenStopped={true}
    size="small"
  />
</View>
`;
